### PR TITLE
Ftp transport2

### PR DESCRIPTION
**Pull request by petervig-srg**

### CODE REVIEW

File: sys/Plugins/Transports/FtpTransport.php [hunk 1]
```diff
@@ -131,9 +141,25 @@ public function get(): array {
...
+		$path = $this->path;
+		if ($path !== '' && substr($path, -1) !== '/') $path .= '/';
+		$contents = [];
+		$newest = '0';
+		foreach ($contentsAll as $entry) {
+			if ($entry['type'] !== 'file') continue;
+			if ($this->filenamePattern !== '' && !fnmatch($this->filenamePattern, $entry['name'])) continue;
+			if (count($contents) > 0 && $entry['modify'] < $newest) continue;
+			if (count($contents) == 0) array_push($contents, $path . $entry['name']);
+			else $contents[0] = $path . $entry['name'];
```
Comment (petervig-srg) [inline sys/Plugins/Transports/FtpTransport.php]: No need to check if the array is empty, you can just push to it.

File: sys/Plugins/Transports/FtpTransport.php [hunk 2]
```diff
@@ -131,9 +141,25 @@ public function get(): array {
...
 
-		$contents = ftp_nlist($conn, $this->path);
+		$contentsAll = ftp_mlsd($conn, $this->path);
+		//die(var_dump($contentsAll));
+
+		// Take only newest of all files which meet the filename pattern
+		$path = $this->path;
+		if ($path !== '' && substr($path, -1) !== '/') $path .= '/';
+		$contents = [];
+		$newest = '0';
```
Comment (petervig-srg) [inline sys/Plugins/Transports/FtpTransport.php]: Let's make this 'newest' mechanic configurable, and I would think not the default behavior.

Comment (petervig-srg) [reply sys/Plugins/Transports/FtpTransport.php]: I would also extract it into its own method, like
```suggestion
		if($this->newestFileOnly)
		$contents = $this->newestFile($contentsAll)
```
Then you also don't need $contentsAll

**Review by petervig-srg — COMMENTED** [review]
Also, please rebase ftp-transport2 onto dev
`git checkout ftp-transport2`
`git rebase dev`
`git push --force`

File: sys/Plugins/Transports/FtpTransport.php (hunk 1 above)
Comment (STASA-DM) [reply sys/Plugins/Transports/FtpTransport.php]: the array should contain only 1 element. So if it already has one, it must be overwritten.

File: sys/Plugins/Transports/FtpTransport.php (hunk 2 above)
Comment (STASA-DM) [reply sys/Plugins/Transports/FtpTransport.php]: OK, I made it configurable and extract the newest mechanic in its own function

**Pull request merged**